File: Notes/CPP/Maths.cpp
// modular exponentiation
ll modPow(ll a, ll b, ll m)
{
    ll res = 1;
    a = a % m;
    while (b)
    {
        if (b & 1)
            res = (res * a) % m;
        a = (a * a) % m;
        b >>= 1;
    }
    return res % m;
}

// modular inverse
ll modInv(ll n, ll p) // if p is prime
{
    return modPow(n, p - 2, p);
}

// gcd
ll gcd(ll a, ll b)
{
    return b ? gcd(b, a % b) : a;
}

// lcm
ll lcm(ll a, ll b)
{
    return a / gcd(a, b) * b;
}

// gcd extended
ll gcdExt(ll a, ll b, ll &x, ll &y)
{
    if (not b)
    {
        x = 1;
        y = 0;
        return a;
    }
    ll x1, y1;
    ll g = gcdExt(b, a % b, x1, y1);
    x = y1;
    y = x1 - y1 * (a / b);
    return g;
}

File: Notes/Python/Maths.py
def modPow(a, b, m):  # modular exponentiation
    a %= m
    res = 1
    while b:
        if b & 1:
            res = (res*a) % m
        a = (a*a) % m
        b >>= 1
    return res % m


def modInv(n, p):  # modular Inverse if p is prime
    return modPow(n, p-2, p)


def gcd(a, b):  # gcd
    if not b:
        return a
    return gcd(b, a % b)


def lcm(a, b):  # lcm
    return a//gcd(a, b) * b


def gcdExt(a, b):  # gcd extended
    if not b:
        return [a, 1, 0]
    g, x1, y1 = gcdExt(b, a % b)
    x = y1
    y = x1-y1*(a//b)
    return [g, x, y]


def modInv(a, m):   # modular inverse general case
    g, x, y = gcdExt(a, m)
    if g != 1:
        return None
    x = ((x % m)+m) % m
    return x
